### PR TITLE
Initial schema for Bundle CRD

### LINKED ISSUE #139: Hot loop on invalid resource format
If a resource in a bundle is misspecified, smith hot loops.
```
E1004 04:09:38.667737       1 streamwatcher.go:109] Unable to decode an event from the watch stream: unable to decode watch event: Object 'Kind' is missing in '.....'
```

### LINKED ISSUE #141: Define OpenAPI schema Bundle CRD
This is necessary to prevent storage of invalid bundles.
https://kubernetes.io/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/#validation

### PR DESCRIPTION
**Pull request by ash2k**
Fixes #141.
Fixes #139.

### CODE REVIEW

File: cmd/smith/app/app.go [hunk 1]
```diff
@@ -207,10 +210,35 @@ func NewFromFlags(flagset *flag.FlagSet, arguments []string) (*App, error) {
 	flagset.IntVar(&a.Workers, "workers", 2, "Number of workers that handle events from informers")
 	flagset.StringVar(&a.Namespace, "namespace", meta_v1.NamespaceAll, "Namespace to use. All namespaces are used if empty string or omitted")
 	pprofAddr := flag.String("pprof-address", "", "Address for pprof to listen on")
+	printBundleSchema := flag.String("print-bundle-schema", "", "Print Bundle schema and exit (specify format: json or yaml)")
```
Comment (nilebox) [inline cmd/smith/app/app.go]: what do you need this application mode for? debugging?

File: docs/deployment/0-crd.yaml [hunk 2]
```diff
@@ -11,3 +11,147 @@ spec:
...
+            apiVersion:
+              $ref: '#/definitions/apiVersion'
+            kind:
+              $ref: '#/definitions/kind'
+          required:
+          - kind
+          - apiVersion
+          type: object
+      properties:
+        spec:
```
Comment (ash2k) [inline docs/deployment/0-crd.yaml]: This schema only validates the `spec` of the `Bundle`. Common metadata is validated by kubernetes itself.

File: cmd/smith/app/app.go [hunk 3]
```diff
@@ -207,10 +210,35 @@ func NewFromFlags(flagset *flag.FlagSet, arguments []string) (*App, error) {
...
+		data, err := yaml.Marshal(resources.BundleCrd().Spec.Validation.OpenAPIV3Schema)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to marshal Bundle schema")
+		}
+		_, err = os.Stdout.Write(data)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed write schema YAML to stdout")
+		}
+		return nil, context.Canceled
+	case "":
```
Comment (nilebox) [inline cmd/smith/app/app.go]: do you actually need a special `case ""` there? isn't `default` enough?
could be some code style thing I guess.

Comment (nilebox) [reply cmd/smith/app/app.go]: if the flag is not set, you should skip printing a bundle schema and run the application I suppose?
as far as I see currently the default value is `""` (empty string), and you will exit with error int that case

Comment (nilebox) [reply cmd/smith/app/app.go]: ah nvm, Go `switch` works differently from other languages (https://tour.golang.org/flowcontrol/9): 
> Go only runs the selected case, not all the cases that follow

can you add some comment there like
```go
case "":
  // flag not set, do nothing
```

File: cmd/smith/app/app.go (hunk 1 above)
Comment (ash2k) [reply cmd/smith/app/app.go]: To get the schema in json or yaml? Because I don't want to write it by hand.

File: cmd/smith/app/app.go (hunk 3 above)
Comment (ash2k) [reply cmd/smith/app/app.go]: Added a comment

File: cmd/smith/app/app.go (hunk 1 above)
Comment (nilebox) [reply cmd/smith/app/app.go]: should be a separate binary then? :)
just a bit confusing given that it's not part of "normal" usage of the application.

Comment (ash2k) [reply cmd/smith/app/app.go]: Yeah, I thought about making it a separate binary. Let me try to see how it'll look...

File: pkg/resources/crd_helpers.go [hunk 4]
```diff
@@ -68,6 +289,7 @@ func EnsureCrdExists(ctx context.Context, defaulter runtime.ObjectDefaulter, cli
 				return nil
 			}
 			log.Printf("Updating CustomResourceDefinition %s", crd.Name)
+			obj = obj.DeepCopy()
```
Comment (ash2k) [inline pkg/resources/crd_helpers.go]: Was mutating original object from lister 🐛

File: cmd/smith/app/app.go [hunk 5]
```diff
@@ -206,7 +206,7 @@ func NewFromFlags(flagset *flag.FlagSet, arguments []string) (*App, error) {
 	flagset.DurationVar(&a.ResyncPeriod, "resync-period", defaultResyncPeriod, "Resync period for informers")
 	flagset.IntVar(&a.Workers, "workers", 2, "Number of workers that handle events from informers")
 	flagset.StringVar(&a.Namespace, "namespace", meta_v1.NamespaceAll, "Namespace to use. All namespaces are used if empty string or omitted")
-	pprofAddr := flag.String("pprof-address", "", "Address for pprof to listen on")
+	pprofAddr := flagset.String("pprof-address", "", "Address for pprof to listen on")
```
Comment (ash2k) [inline cmd/smith/app/app.go]: 🐛

File: cmd/smith/app/app.go (hunk 1 above)
Comment (ash2k) [reply cmd/smith/app/app.go]: Looks better now!

File: pkg/resources/crd_helpers.go [hunk 6]
```diff
@@ -32,10 +38,225 @@ func BundleCrd() *apiext_v1b1.CustomResourceDefinition {
 				Singular: smith_v1.BundleResourceSingular,
 				Kind:     smith_v1.BundleResourceKind,
 			},
+			Validation: &apiext_v1b1.CustomResourceValidation{
+				OpenAPIV3Schema: &apiext_v1b1.JSONSchemaProps{
```
Comment (wryun) [inline pkg/resources/crd_helpers.go]: soo... typey?

File: docs/deployment/0-crd.yaml [hunk 7]
```diff
@@ -11,3 +11,147 @@ spec:
     singular: bundle
   scope: Namespaced
   version: v1
+  validation:
```
Comment (wryun) [inline docs/deployment/0-crd.yaml]: why do we have 0-crd.yaml when Smith creates it automatically?

Comment (ash2k) [reply docs/deployment/0-crd.yaml]: It does not anymore https://github.com/atlassian/smith/pull/184

**Pull request merged**